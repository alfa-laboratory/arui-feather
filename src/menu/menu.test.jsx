/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from 'react';
import { mount, shallow } from 'enzyme';

import Menu from './menu';

const MENU_ITEM1 = {
    type: 'item',
    content: 'MenuItem 1',
    value: 'value1',
    props: {
        url: '#1'
    }
};

const MENU_ITEM2 = {
    type: 'item',
    content: 'MenuItem 2',
    value: 'value2',
    props: {
        url: '#2'
    }
};

const MENU_ITEM2_CLONE = {
    type: 'item',
    content: 'MenuItem 2',
    value: 'value2',
    key: 'value3'
};

const MENU_GROUP = {
    type: 'group',
    title: 'Group Title',
    content: [
        MENU_ITEM1
    ]
};

function getNumberOfItemsWithClass(items, className) {
    return items.reduce((prev, item) => {
        if (item.getDOMNode().className.indexOf(className) !== -1) {
            return prev + 1;
        }
        return prev;
    }, 0);
}

describe('menu', () => {
    beforeEach(() => {
        jest.useFakeTimers();
    });

    afterEach(() => {
        jest.useRealTimers();
    });

    it('should render without problem', () => {
        let menu = mount(<Menu />);

        expect(menu).toMatchSnapshot();
    });

    it('should render menu items as type=`block` when property `mode` is identified', () => {
        let menu = mount(<Menu mode='check' content={ [MENU_ITEM1, MENU_ITEM2] } />);
        let menuFirstItemNode = menu.find('.menu-item').at(0);

        expect(menuFirstItemNode.getDOMNode().className).toContain('menu-item_type_block');
    });

    it('should render menu items in group with group title', () => {
        let menu = mount(<Menu content={ [MENU_GROUP] } />);

        expect(menu).toMatchSnapshot();
    });

    it('should auto check first item when mode=`radio` and `checkedItems` property is empty', () => {
        let menu = mount(<Menu mode='radio' content={ [MENU_ITEM1, MENU_ITEM2] } />);
        let menuFirstItemNode = menu.find('.menu-item').at(0);

        expect(menuFirstItemNode.getDOMNode().className).toContain('menu-item_checked');
    });

    it('shouldn\'t crash when mode=`radio`, and `checkedItems` and `content` properties is empty', () => {
        let menu = mount(<Menu mode='radio' content={ [] } />);

        expect(menu).toMatchSnapshot();
    });

    it('should have just one checked menu-item when mode=`radio`', () => {
        let menu = mount(<Menu mode='radio' content={ [MENU_ITEM1, MENU_ITEM2] } />);
        let menuChildsNode = menu.find('.menu-item');

        menuChildsNode.forEach(child => child.simulate('click'));

        let checkedMenuItems = menu.find('.menu-item_checked');
        expect(checkedMenuItems.length).toEqual(1);
    });

    it('should check every menu-item when mode=`check`', () => {
        let menu = mount(<Menu mode='check' content={ [MENU_ITEM1, MENU_ITEM2] } />);
        let menuChildsNode = menu.find('.menu-item');

        menuChildsNode.forEach(child => child.simulate('click'));

        let checkedMenuItems = menu.find('.menu-item_checked');
        expect(checkedMenuItems.length).toEqual(menuChildsNode.length);
    });

    it('should disabled all menu-items when menu property disabled=true', () => {
        let menu = mount(<Menu disabled={ true } content={ [MENU_ITEM1, MENU_ITEM2] } />);
        let menuChildsNode = menu.find('.menu-item');
        let menuItemsNum = getNumberOfItemsWithClass(menuChildsNode, 'menu-item_disabled');

        expect(menuItemsNum).toBe(menuChildsNode.length);
    });

    it('should render without problem when give item with duplicate value', () => {
        let content = [MENU_ITEM1, MENU_ITEM2, MENU_ITEM2_CLONE];
        let menu = mount(<Menu content={ content } />);
        let menuChildNodes = menu.find('.menu-item');

        expect(menuChildNodes.length).toBe(content.length);
    });

    it('should call `onItemCheck` callback after menu-item was clicked and if `mode` is identified', () => {
        let onItemCheck = jest.fn();
        let menu = mount(<Menu onItemCheck={ onItemCheck } mode='check' content={ [MENU_ITEM1, MENU_ITEM2] } />);

        menu.find('.menu-item').first().simulate('click');

        expect(onItemCheck).toHaveBeenCalled();
    });

    it('should call `onMouseEnter` callback after menu was hovered', () => {
        let onMouseEnter = jest.fn();
        let menu = shallow(<Menu onMouseEnter={ onMouseEnter } />);

        menu.simulate('mouseEnter');

        expect(onMouseEnter).toHaveBeenCalled();
    });

    it('should call `onMouseLeave` callback after menu was unhovered', () => {
        let onMouseLeave = jest.fn();
        let menu = shallow(<Menu onMouseLeave={ onMouseLeave } />);

        menu.simulate('mouseLeave');

        expect(onMouseLeave).toHaveBeenCalled();
    });

    it('should set/unset class when menu-item hovered/unhovered', () => {
        let menu = shallow(<Menu />);

        menu.simulate('mouseEnter');
        expect(menu.hasClass('menu_hovered')).toBeTruthy();

        menu.simulate('mouseLeave');
        expect(menu.hasClass('menu_hovered')).toBeFalsy();
    });

    it('should focus on menu after `focus` method call', () => {
        let menu = mount(<Menu />);

        menu.instance().focus();

        let control = menu.instance().root;
        jest.spyOn(control, 'focus');

        menu.instance().focus();

        expect(control.focus).toHaveBeenCalled();
    });

    it('should remove focus from menu after `blur` method call', () => {
        // we need to have fake body to get document.activeElement
        document.body.innerHTML = '<button id="btn1">btn 1 </button>';
        document.getElementById('btn1').focus();

        let menu = mount(<Menu />);

        document.activeElement.blur = jest.fn();

        menu.instance().blur();

        expect(document.activeElement.blur).toHaveBeenCalled();
    });

    it('should call `onFocus` callback after menu was focused', () => {
        let onFocus = jest.fn();
        let menu = mount(<Menu onFocus={ onFocus } />);

        menu.simulate('focus');

        expect(onFocus).toHaveBeenCalled();
    });

    it('should call `onBlur` callback after menu lost focus', () => {
        let onBlur = jest.fn();
        let menu = mount(<Menu onBlur={ onBlur } />);

        menu.simulate('blur');

        jest.runAllTimers();
        expect(onBlur).toHaveBeenCalled();
    });

    it('should return root `HTMLElement` after `getNode` method call', () => {
        let menu = mount(<Menu />);

        let node = menu.instance().getNode();

        expect(node).toBeInstanceOf(HTMLElement);
    });
});
