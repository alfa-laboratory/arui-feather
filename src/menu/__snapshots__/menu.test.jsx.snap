// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu should render menu items in group with group title 1`] = `
<Menu
  autoFocusFirstItem={false}
  content={
    Array [
      Object {
        "content": Array [
          Object {
            "content": "MenuItem 1",
            "props": Object {
              "url": "#1",
            },
            "type": "item",
            "value": "value1",
          },
        ],
        "title": "Group Title",
        "type": "group",
      },
    ]
  }
  groupView="default"
  mode="basic"
  size="m"
>
  <div
    className="menu menu_size_m menu_group-view_default menu_mode_basic menu_theme_alfa-on-white"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    tabIndex={0}
  >
    <div
      className="menu__group"
      key="group_0"
    >
      <div
        className="menu__group-title"
      >
        Group Title
      </div>
      <MenuItem
        checked={false}
        hovered={null}
        key="value1"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        size="m"
        type="link"
        url="#1"
        value="value1"
        view="plain"
      >
        <div
          className="menu-item menu-item_type_link menu-item_theme_alfa-on-white"
        >
          <Link
            checked={false}
            className="menu-item__control menu-item__link"
            disabled={false}
            iconPosition="left"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            pseudo={false}
            size="m"
            tabIndex={0}
            text={false}
            url="#1"
          >
            <a
              className="link link_size_m link_theme_alfa-on-white menu-item__control menu-item__link"
              href="#1"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              tabIndex={0}
            >
              MenuItem 1
            </a>
          </Link>
        </div>
      </MenuItem>
    </div>
  </div>
</Menu>
`;

exports[`menu should render without problem 1`] = `
<Menu
  autoFocusFirstItem={false}
  groupView="default"
  mode="basic"
  size="m"
>
  <div
    className="menu menu_size_m menu_group-view_default menu_mode_basic menu_theme_alfa-on-white"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    tabIndex={0}
  />
</Menu>
`;

exports[`menu shouldn't crash when mode=\`radio\`, and \`checkedItems\` and \`content\` properties is empty 1`] = `
<Menu
  autoFocusFirstItem={false}
  content={Array []}
  groupView="default"
  mode="radio"
  size="m"
>
  <div
    className="menu menu_size_m menu_group-view_default menu_mode_radio menu_theme_alfa-on-white"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    tabIndex={0}
  />
</Menu>
`;
